### ISSUE TITLE
`docker compose up --build` not recreating container

### ISSUE THREAD

**Issue by stefanadelbert** [docker/compose#9308]
# Description

When making changes to something that affects an image that a service uses, running `docker compose up --build` should rebuild that image and then recreate the corresponding service's container so that it uses the new image.

> If there are existing containers for a service, and the service’s configuration or image was changed after the container’s creation, docker-compose up picks up the changes by stopping and recreating the containers (preserving mounted volumes). To prevent Compose from picking up changes, use the --no-recreate flag.
See https://docs.docker.com/compose/reference/up/.

I've noticed recently that this is no longer the case, i.e. if I make a change to a dependency of an image, `docker compose up --build` will rebuild the image (as expected), but will not recreate the container. `docker image ls -a` shows both the old and new images, but `docker compose images` shows that the old images is still being used.

This behaviour I'm seeing seems similar to the following issue:
- https://github.com/docker/compose-cli/issues/1608

# Steps to reproduce the issue
## Minimal files

`compose.yaml`
```yaml
version: "3.7"

services:
  a:
    build:
      context: .
```
`Dockerfile`
```dockerfile
FROM python:3.8-slim

COPY main.py ./main.py
CMD python main.py
```
`main.py`
```python
print("1")
```

## Steps

- `docker compose up --build`
- `docker compose images`
    - note the `Image Id`
- modify `main.py`
  - change `print('1')` to `print('2')`
- `docker compose up --build`
    - a new image will be created because a dependency (`main.py`) has changed
    - however the **old** code in `main.py` executes
- `docker compose images`
    - the **old** `Image Id` is still listed
- `docker compose up --build --force-recreate`
    - the modified code in `main.py` executes
- `docker compose images`
    - the new `Image Id` is now listed

# Additional information you deem important
This started happening recently, after a system upgrade.

# Output
## Output of `docker compose version`

```shell
Docker version 20.10.12, build e91ed5707e
Docker Compose version 2.3.3
```

## Output of `docker info`

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.7.1-docker)
  compose: Docker Compose (Docker Inc., 2.3.3)

Server:
 Containers: 6
  Running: 4
  Paused: 0
  Stopped: 2
 Images: 171
 Server Version: 20.10.12
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: false
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: uq13g9060lepbcwdmdziw6zbz
  Is Manager: true
  ClusterID: sj5b54ikcmgj5dkclf8hvj1y8
  Managers: 1
  Nodes: 1
  Default Address Pool: 10.0.0.0/8
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 10.10.10.137
  Manager Addresses:
   10.10.10.137:2377
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 10f428dac7cec44c864e1b830a4623af27a9fc70.m
 runc version: v1.1.0-0-g067aaf85
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.15.28-1-MANJARO
 Operating System: Manjaro Linux
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 31.34GiB
 Name: stefan-desktop
 ID: XRDU:VYJ7:LNKD:HLJT:6PNO:BL6A:XCPC:WE6O:7RI6:HPFQ:5SSV:3NAR
 Docker Root Dir: /mnt/media/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

# Additional environment details
None

**Comment by jensneuhaus** [comment]
I can confirm this problem as I have the same issue and I could also reproduce it with the given minimal example!

It worked fine before and stopped working probably after a `brew update` yesterday.


### My output from `docker info`

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.8.1)
  compose: Docker Compose (Docker Inc., v2.3.3)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 22
  Running: 13
  Paused: 0
  Stopped: 9
 Images: 56
 Server Version: 20.10.13
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 2a1d4dbdb2a1030dc5b01e96fb110a9d9f150ecc
 runc version: v1.0.3-0-gf46b6ba
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.10.104-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 4
 Total Memory: 31.31GiB
 Name: docker-desktop
 ID: ...
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5000
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by glours** [comment]
We just released the `v2.3.4` version which should fix this issue, can you test it and close this ticket if it's ok for you please?

**Comment by stefanadelbert** [comment]
I can confirm that the v2.3.4 fixes the behaviour.
